### PR TITLE
Update loss function for jskit::BiEncoder 

### PR DESCRIPTION
**Pull request by StefanHeng**
Update: efficient dot product loss; New: cosine similarity loss

See [issue 145](https://github.com/Jaseci-Labs/jaseci/issues/145).

### CODE REVIEW

File: jskit/encoders/utils/models.py [hunk 1]
```diff
@@ -224,11 +226,14 @@ def forward(self, context_input_ids=None, context_input_masks=None,
...
-            mask = torch.eye(context_input_ids.size(0)).to(
-                context_input_ids.device)
-            loss = F.log_softmax(dot_product, dim=-1) * mask
+            if self.loss_type == 'dot-prod':
+                dot_product = torch.sum(context_vec * candidate_vec, dim=-1)  # row-wise dot product
+                loss = F.log_softmax(dot_product, dim=-1)
+            else:  # `cos-sim`
+                loss_fn = torch.nn.CosineEmbeddingLoss()
+                batch_size = candidate_vec.size(0)
+                # TODO: support negative pairs
```
Comment (ypkang) [inline jskit/encoders/utils/models.py:235]: So this doesn't work with negative examples?

**Comment by ypkang** [comment]
The changes look good. @AshishMahendra could you take a look here too and give it a test?

**Comment by AshishMahendra** [comment]
@StefanHeng did you test for the loss and accuracy, after implementing ?

**Review by AshishMahendra — COMMENTED** [review]
Seems like loss value is not reducing and accuracy score have gone down

File: jskit/encoders/utils/models.py (hunk 1 above)
Comment (StefanHeng) [reply jskit/encoders/utils/models.py:235]: https://github.com/Jaseci-Labs/jaseci/pull/147/commits/cf7c5afeb8828d9a20cbdbc56ab769e6a78d88cf#diff-eef0e0493716527b9006d448f729d8c4ff5f4a52988f05999bf95907d0e0e134R236

Here we assign only positive labels as the target. With respect to loss we can have it work with negative examples by just matching the labels from the batch. However, supporting negative examples requires more changes in the entire forward pass. 

So no it doesn't work with negative examples now.

File: jskit/encoders/utils/models.py [hunk 2]
```diff
@@ -224,11 +226,14 @@ def forward(self, context_input_ids=None, context_input_masks=None,
                     return candidate_vec
         if labels is not None and mode == "train":
             candidate_vec = candidate_vec.squeeze(1)
-            dot_product = torch.matmul(
-                context_vec, candidate_vec.t())  # [bs, bs]
-            mask = torch.eye(context_input_ids.size(0)).to(
-                context_input_ids.device)
-            loss = F.log_softmax(dot_product, dim=-1) * mask
+            if self.loss_type == 'dot-prod':
```
Comment (ypkang) [inline jskit/encoders/utils/models.py:229]: @StefanHeng any reason why you also modified the dot product code as well?

Comment (StefanHeng) [reply jskit/encoders/utils/models.py:229]: IRRC as we discussed last time and I mentioned in the issue 145, we’re effectively computing the dot product for each corresponding pair of context and candidate vectors. I made the change cos I believe it’s more efficient

**Comment by StefanHeng** [comment]
I could take a look when I'm free (probably after I graduate). I'll need to run experiments to see why the drop in performance. 

Per our meeting with @marsninja yesterday, I want to focus on the paper (and schoolwork) for now. We'll get into trying loss functions.

**Comment by ypkang** [comment]
> I could take a look when I'm free (probably after I graduate). I'll need to run experiments to see why the drop in performance.
> 
> Per our meeting with @marsninja yesterday, I want to focus on the paper (and schoolwork) for now. We'll get into trying loss functions.

Sounds good. @AshishMahendra is going to take this over and incorporate the changes.

**Comment by ypkang** [comment]
This is now superseded by #205
